src/bit-fields/src/bit_range.rs

```diff
@@ -0,0 +1,1022 @@
+            #[inline]
+            pub fn checked_add_assign(
+                &mut self,
+                x: $y,
+            ) -> Result<(), $crate::CheckedAddAssignError> {
+                if x <= Self::MAX {
+                    let cur = (BitRangeMut::<$x, START, END>::MASK & self.0.get()) >> START;
+                    if x.get() <= BitRangeMut::<$x, START, END>::MAX - cur {
+                        let shift = x.get() << START;
+                        // SAFETY: Always safe.
+                        *self.0 = unsafe { <$y>::new_unchecked(self.0.get() + shift) };
+                        Ok(())
+                    } else {
+                        Err($crate::CheckedAddAssignError::Overflow)
+                    }
+                } else {
+                    Err($crate::CheckedAddAssignError::OutOfRange)
+                }
+            }
+
+            // `x <= Self::MAX` guarantees `x << START` is safe and `x <= cur` guarantees
+            // `self.data_mut() -= shift` is safe.
+            #[doc = concat!("
+                Subtract `x` from the value of the bit range.
+
+                # Errors
+
+                1. When `x` is greater than the maximum value storable in the bit range.
+                2. When subtracting `x` from the value of the bit range would underflow.
+
+                ```
+                use bit_fields::{BitRangeMut, CheckedSubAssignError};
+                use std::num::", stringify!($y) ,";
+                let mut x = ", stringify!($y), "::new(18).unwrap();
+
+                let mut nibble = BitRangeMut::<_,0,4>(&mut x);
+                assert_eq!(nibble,", stringify!($y), "::new(2).unwrap());
+                unsafe {
+                    assert_eq!(nibble.checked_sub_assign(", stringify!($y), "::new(16).unwrap()),Err(CheckedSubAssignError::OutOfRange));
+                    assert_eq!(nibble.checked_sub_assign(", stringify!($y), "::new(3).unwrap()),Err(CheckedSubAssignError::Underflow));
+                    assert_eq!(nibble.checked_sub_assign(", stringify!($y), "::new(1).unwrap()),Ok(()));
+                }
+                assert_eq!(nibble,", stringify!($y), "::new(1).unwrap());
+                assert_eq!(x,", stringify!($y), "::new(17).unwrap());
+
+                let mut nibble = BitRangeMut::<_,4,8>(&mut x);
+                assert_eq!(nibble,", stringify!($y), "::new(1).unwrap());
+                unsafe {
+                    assert_eq!(nibble.checked_sub_assign(", stringify!($y), "::new(16).unwrap()),Err(CheckedSubAssignError::OutOfRange));
+                    assert_eq!(nibble.checked_sub_assign(", stringify!($y), "::new(2).unwrap()),Err(CheckedSubAssignError::Underflow));
+                    assert_eq!(nibble.checked_sub_assign(", stringify!($y), "::new(1).unwrap()),Ok(()));
+                }
+                
+                unsafe {
+                    assert_eq!(nibble,", stringify!($y), "::new_unchecked(0));
+                }
```

Comment:
Miri (rightfully) complains about this test hitting undefined behavior.

Maybe `checked_sub_assign` should treat the "would result in 0" case as an error, instead of being unsafe? I think returning `Ok(())` for something that triggers undefined behavior is a bit misleading.

Reply:
Rewrote the non-zero `checked_sub_assign`:
```rust
#[allow(clippy::integer_arithmetic, clippy::arithmetic_side_effects)]
#[inline]
pub fn checked_sub_assign(
    &mut self,
    x: $x,
) -> Result<(), $crate::CheckedSubAssignError> {
    if x <= Self::MAX {
        let cur = (Self::MASK & *self.0) >> START;
        if x <= cur {
            let shift = x << START;
            *self.0 -= shift;
            Ok(())
        } else {
            Err($crate::CheckedSubAssignError::Underflow)
        }
    } else {
        Err($crate::CheckedSubAssignError::OutOfRange)
    }
}
```